woocommerce/grow
# Grow Smooth Generator

### Changes proposed in this Pull Request:

This was a hack day project to create a plugin which can generate data for extensions that are maintained by Grow. The plugin is based off of the [WooCommerce Smooth Generator](https://github.com/woocommerce/wc-smooth-generator#readme) which must be activated on the site.

Initially I focused on setting up a basic scaffold for the plugin and then started with adding commands for generating AutomateWoo data. To limit the scope of this project I only implemented WP CLI commands. I was able to add the following two generators for AutomateWoo:

- Conversion data
- Email and SMS tracking data (opens and clicks)

I didn't get round to including any build process or packaging the plugins into a zip file. So for now the install must be done manually.

### Example output:
```
$ wp wc generate aw-email-tracking 10 --workflow=142 --date-start=2023-04-01 --date-end=2023-04-10
Initializing...
Generating email and SMS tracking  100% [================================================================================] 0:00 / 0:00
Success: 10 email and SMS tracking generated in 0.05 seconds
```

### Detailed test instructions:
1. As prerequisites make sure both the [WooCommerce Smooth Generator](https://github.com/woocommerce/wc-smooth-generator) plugin and AutomateWoo are installed and activated on the site
2. Clone the grow repo locally 
3. Create a symbolic link in your WordPress plugins directory `wp-content/plugins` to the folder `plugins/grow-smooth-generator`
4. In the plugin folder run `composer install`
5. Activate the plugin `wp plugin activate grow-smooth-generator`
6. Generate a workflow in AutomateWoo and make sure to check "Enable tracking" and "Enable conversion tracking"
![image](https://user-images.githubusercontent.com/11388669/233420295-763e7661-d3d1-4d9c-8922-bcf7b7082362.png)
7. Run a WP CLI command to generate some conversion data `wp wc generate aw-conversions 10 --workflow=<workflow ID> --date-start=2023-04-01 --date-end=2023-04-10`
8. Go to Analytics > Conversions and select a report for the month of April, confirm that we can see the generated conversion data from the 1st till the 10th of April:
![image](https://user-images.githubusercontent.com/11388669/233421106-fc574670-902e-4838-bee8-c6fe9d8edc82.png)
9. Run a WP CLI command to generate email tracking data `wp wc generate aw-email-tracking 10 --workflow=<workflow ID> --date-start=2023-04-01 --date-end=2023-04-10`
10. Go to Analytics > Email & SMS Tracking and select a report for the month of April, confirm that we can see the generated opens/clicks from the 1st till the 10th of April:
![image](https://user-images.githubusercontent.com/11388669/233426507-86a6a54f-cfcf-4c47-a368-7b5e60dcc122.png)

> **Note**
If you viewed the report for the month of April (causing the data to be cached), and then generated the new data, then it might not show up right away in the results. This is because WC Analytics doesn't have a way to clear the cache for a set of reports, we have an open issue for this here: https://github.com/woocommerce/automatewoo/issues/1355
To get around this you can clear the cache manually or select a different date range in the reports.

### Additional Information
The plugin has been structured to easily add additional commands for AutomateWoo as well as add generators for other extensions.

See `includes/CLI.php` for adding an additional extension, an additional class must also be created in `includes/CLI` for registering the extensions commands.
Individual generators can be created in `includes/Generator` and can be group in a sub folder for each extension.

### Changelog entry
* Add - Create a Grow Smooth Generator plugin.
